master/internal/prom/prom.go

```diff
@@ -0,0 +1,233 @@
+	allocationIDToTaskID.WithLabelValues(aID, tID).Dec()
+}
+
+// AssociateContainerRuntimeID associates a Determined container ID with the docker container ID.
+func AssociateContainerRuntimeID(cID string, dcID string) {
+	containerIDToRuntimeID.WithLabelValues(dcID, cID).Inc()
+}
+
+// AddAllocationContainer associates allocation and container and container and GPUs.
+func AddAllocationContainer(summary sproto.ContainerSummary) {
+	AssociateAllocationContainer(summary.AllocationID.String(), summary.ID.String())
+	for _, d := range summary.Devices {
+		AssociateContainerGPU(summary.ID.String(), d)
+	}
+}
+
+// RemoveAllocationContainer disassociates allocation and container and container and its GPUs.
+func RemoveAllocationContainer(summary sproto.ContainerSummary) {
+	DisassociateAllocationContainer(summary.AllocationID.String(), summary.ID.String())
+	for _, d := range summary.Devices {
+		DisassociateContainerGPU(summary.ID.String(), d)
+	}
+}
+
+// DisassociateAllocationContainer disassociates allocation ID with its container ID.
+func DisassociateAllocationContainer(aID string, cID string) {
+	containerIDToAllocationID.WithLabelValues(cID, aID).Dec()
+}
+
+// AssociateExperimentIDLabels assicates experiment ID with a list of labels.
+func AssociateExperimentIDLabels(eID string, labels []string) {
+	for i := range labels {
+		experimentIDToLabels.WithLabelValues(eID, labels[i]).Inc()
+	}
+}
+
+// AssociateContainerGPU associates container ID with GPU device ID.
+func AssociateContainerGPU(cID string, d device.Device) {
+	if d.Type == device.GPU {
+		gpuUUIDToContainerID.
+			WithLabelValues(d.UUID, cID).
+			Inc()
+	}
+}
+
+// DisassociateContainerGPU removes association between container ID and device ID.
+func DisassociateContainerGPU(cID string, d device.Device) {
+	if d.Type != device.GPU {
+		return
+	}
+
+	gpuUUIDToContainerID.WithLabelValues(d.UUID, cID).Dec()
+	gpuUUIDToContainerID.DeleteLabelValues(d.UUID, cID)
+}
+
+// AddAgentAsTarget adds an entry to a list of currently active agents in a target JSON file.
+// This file is watched by Prometheus for changes to which targets to scrape.
+func AddAgentAsTarget(
+	ctx *actor.Context,
+	agentID string,
+	agentAddress string,
+	agentResourcePool string) {
+	ctx.Log().Infof("adding agent %s at %s as a prometheus target", agentID, agentAddress)
+
+	if _, err := os.Stat(targetsFile); os.IsNotExist(err) {
+		_, err = os.OpenFile(targetsFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
+		if err != nil {
+			ctx.Log().Errorf("Error creating targets config file %s", err)
+		}
+	}
+
+	fileSDConfig := fileSDConfigEntry{
+		Targets: []string{
+			agentAddress + dcgmPort,
+			agentAddress + cAdvisorPort,
+		}, Labels: map[string]string{
+			detAgentIDLabel:      agentID,
+			detResourcePoolLabel: agentResourcePool,
+		},
+	}
+
+	fileSDConfigs := getFileSDConfigs()
+
+	fileSDConfigs = append(fileSDConfigs, fileSDConfig)
+
+	err := writeConfigsToTargetsFile(fileSDConfigs)
```

Comment:
This is interesting and works for now; maybe though it is better to expose an api like `/prom/det-file-sd-config` that returns the file sd config. I think that would be more flexible for users to consume.

Reply:
either way though, I think this file would need a mutex?